Source/WebCore/page/scrolling/ScrollAnchoringController.h

```diff
@@ -0,0 +1,52 @@
+/*
+* Copyright (C) 2022 Apple Inc. All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions
+* are met:
+* 1. Redistributions of source code must retain the above copyright
+*    notice, this list of conditions and the following disclaimer.
+* 2. Redistributions in binary form must reproduce the above copyright
+*    notice, this list of conditions and the following disclaimer in the
+*    documentation and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+* PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+* BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+* THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#pragma once
+
+#include "FloatPoint.h"
+#include "FrameView.h"
+
+namespace WebCore {
+
+class Element;
+
+class ScrollAnchoringController final {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit ScrollAnchoringController(FrameView& frame)
+        : m_frame(frame)
+    { }
+    void invalidateAnchorNode();
+    void updateScrollPosition();
+    void selectAnchorNode();
+    FrameView& frameView() { return m_frame; }
+
+private:
+    FrameView& m_frame;
+    WeakPtr<Element> m_anchorElement;
+    FloatPoint m_lastPositionForAnchorNode;
```

Comment:
Now we have element/node confusion. The spec calls it "node" but in reality I think it has to be an Element (Nodes include Text objects which don't create their own Boxes); the CSS spec says "Each box in the [box tree](https://drafts.csswg.org/css-display-3/#box-tree) represents its corresponding [element](https://drafts.csswg.org/css-display-3/#elements)" [https://drafts.csswg.org/css-display-3/#box].

I think it would be best to call it "Element" everywhere, including in your stub methods.

Reply:
Agreed. Fixed.